src/DynamicExpresso.Core/ParseResult.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System.Collections.Generic;
+using System.Linq.Expressions;
+
+namespace DynamicExpresso
+{
+	/// <summary>
+	/// Represents an expression parse result.
+	/// </summary>
+	public class ParseResult
+	{
+		public ParseResult(
+			Expression expression,
+			IEnumerable<ParameterExpression> usedParameters,
+			IEnumerable<ParameterExpression> declaredParameters,
+			IEnumerable<ReferenceType> types,
+			IEnumerable<Identifier> identifiers)
+		{
+			Expression = expression;
+			UsedParameters = usedParameters;
+			DeclaredParameters = declaredParameters;
+			Types = types;
+			Identifiers = identifiers;
+		}
+
+		/// <summary>
+		/// Gets the parsed expression.
+		/// </summary>
+		/// <value>The expression.</value>
+		public Expression Expression { get; }
+
+		/// <summary>
+		/// Gets the parameters actually used in the expression parsed.
+		/// </summary>
+		/// <value>The used parameters.</value>
+		public IEnumerable<ParameterExpression> UsedParameters { get; }
+
+		/// <summary>
+		/// Gets the parameters declared when parsing the expression.
+		/// </summary>
+		/// <value>The declared parameters.</value>
+		public IEnumerable<ParameterExpression> DeclaredParameters { get; }
+
+		/// <summary>
+		/// Gets the references types in parsed expression.
+		/// </summary>
+		/// <value>The references types.</value>
+		public IEnumerable<ReferenceType> Types { get; }
+
+		/// <summary>
+		/// Gets the identifiers in parsed expression.
+		/// </summary>
+		/// <value>The identifiers.</value>
+		public IEnumerable<Identifier> Identifiers { get; }
+	}
+
+	public class ParseResult<TDelegate> : ParseResult
+	{
+		public ParseResult(
+			Expression expression,
+			IEnumerable<ParameterExpression> usedParameters,
+			IEnumerable<ParameterExpression> declaredParameters,
+			IEnumerable<ReferenceType> types,
+			IEnumerable<Identifier> identifiers) : base(expression, usedParameters, declaredParameters, types, identifiers)
+		{
+		}
+	}
```

Comment:
Why do you have a generic version of `ParseResult`? It seems you don't use the generic parameter?

Reply:
This generic ParseResult<> will store Type information and there is a generic Compile extension over ParseResult<> to Compile to target delegate.

You could perhaps Compile over basic ParseResult, but the intention was to Parse with proper ReturnType and then just compile to get current delegate without passing a DelegateType.

Reply:
Ah, I see it now, thanks for the explanation. Good idea 👍